server/routes/warningDetails.ts

```diff
@@ -1,10 +1,287 @@
+
+  post('/warning-details/:id', async (req, res, next) => {
+    const { id } = req.params
+    const token = await hmppsAuthClient.getSystemClientToken(res.locals.user.username)
+    const breachNoticeApiClient = new BreachNoticeApiClient(token)
+    const ndeliusIntegrationApiClient = new NdeliusIntegrationApiClient(token)
+    const breachNotice = await breachNoticeApiClient.getBreachNoticeById(id as string)
+
+    if (await commonUtils.redirectRequired(breachNotice, res)) return
+
+    const warningDetails = await ndeliusIntegrationApiClient.getWarningDetails(breachNotice.crn, breachNotice.id)
+
+    // failures recorded on this order
+    // list of contacts
+    const contactList: BreachNoticeContact[] = []
+    // select the failures being enforced
+    // list of requirements
+    const requirementList: BreachNoticeRequirement[] = []
+
+    // for final warning theres only 1 of these for first warning screen
+    const failureRecordedContactId: string = req.body.failureRecordedContact
+
+    // lookup the contact and push to the list of contacts to be saved
+    const enforceableContact: EnforceableContact = warningDetails.enforceableContacts.find(
+      contact => contact.id.toString() === failureRecordedContactId,
+    )
+
+    if (enforceableContact) {
+      contactList.push({
+        id: null,
+        contactId: enforceableContact.id,
+        breachNoticeId: breachNotice.id,
+        contactDate: enforceableContact.datetime.toString(),
+        contactType: enforceableContact.type.description,
+        contactOutcome: enforceableContact.outcome.description,
+      })
+    }
+
+    // add the list of contacts to out breach notice
+    breachNotice.breachNoticeContactList = contactList
+
+    // lookup the requirements
+    // this can come in an array or singular
+    const selectedFailureRequirements = asArray(req.body.failuresBeingEnforcedRequirements)
+
+    selectedFailureRequirements.forEach((requirementId: string) => {
+      const enforceableContactWithRequirement: EnforceableContact = warningDetails.enforceableContacts.find(
+        contact => contact.requirement.id.toString() === requirementId,
+      )
+      const bodyParamBreachReason: string = `breachreason${requirementId}`
+
+      const currentRequirement: BreachNoticeRequirement = {
+        id: null,
+        breachNoticeId: breachNotice.id,
+        requirementId: enforceableContactWithRequirement.requirement.id,
+        requirementTypeMainCategoryDescription: enforceableContactWithRequirement.requirement.type.description,
+        requirementTypeSubCategoryDescription: enforceableContactWithRequirement.requirement.subType.description,
+        rejectionReason: req.body[bodyParamBreachReason],
+        fromDate: null,
+        toDate: null,
+      }
+
+      requirementList.push(applyContactDateRangesToRequirement(currentRequirement, warningDetails.enforceableContacts))
+    })
+
+    breachNotice.breachNoticeRequirementList = requirementList
```

Comment:
You could simplify this a bit, using map instead of forEach:
```suggestion
    breachNotice.breachNoticeRequirementList = asArray(req.body.failuresBeingEnforcedRequirements).map((requirementId: string) => {
      const enforceableContactWithRequirement: EnforceableContact = warningDetails.enforceableContacts.find(
        contact => contact.requirement.id.toString() === requirementId,
      )
      const bodyParamBreachReason: string = `breachreason${requirementId}`
      const currentRequirement: BreachNoticeRequirement = {
        id: null,
        breachNoticeId: breachNotice.id,
        requirementId: enforceableContactWithRequirement.requirement.id,
        requirementTypeMainCategoryDescription: enforceableContactWithRequirement.requirement.type.description,
        requirementTypeSubCategoryDescription: enforceableContactWithRequirement.requirement.subType.description,
        rejectionReason: req.body[bodyParamBreachReason],
        fromDate: null,
        toDate: null,
      }

      return applyContactDateRangesToRequirement(currentRequirement, warningDetails.enforceableContacts))
    })
```

Reply:
fixed